File: sdk/python/angorapay.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional
from urllib.parse import urlencode

import requests

__version__ = "0.1.0"


@dataclass
class AngoraPay:
    base_url: str
    api_key: Optional[str] = None
    timeout: int = 30

    def _headers(self) -> Dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        return headers

    def _request(self, method: str, path: str, **kwargs: Any) -> Dict[str, Any]:
        response = requests.request(
            method,
            f"{self.base_url.rstrip('/')}{path}",
            headers=self._headers(),
            timeout=self.timeout,
            **kwargs,
        )
        try:
            payload = response.json()
        except Exception:
            payload = {"raw": response.text}
        if response.status_code >= 400:
            raise RuntimeError(payload.get("error") or f"Angora request failed with {response.status_code}")
        return payload

    def create_mission(self, **payload: Any) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/missions", json=payload)

    def discover_market_services(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/services/search?{urlencode(params)}")

    def simulate_route(self, mission_id: Optional[str] = None, max_price: str = "0.01") -> Dict[str, Any]:
        params = {"max_price": max_price}
        if mission_id:
            params["mission_id"] = mission_id
        return self._request("GET", f"/v1/angora/route/simulate?{urlencode(params)}")

    def call_market_service(self, **payload: Any) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/gateway/call", json=payload)

    def run_demo_market_mission(self, payload: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/demo/market-mission", json={"payload": payload or {}})

    def record_user(self, **payload: Any) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/traction/users", json=payload)

    def record_feedback(self, **payload: Any) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/traction/feedback", json=payload)

    def execution_history(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/execution-history?{urlencode(params)}")

    def submission_metrics(self) -> Dict[str, Any]:
        return self._request("GET", "/v1/angora/submission/metrics")

    def production_readiness(self) -> Dict[str, Any]:
        return self._request("GET", "/v1/angora/production/readiness")

    def receipts(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/receipts?{urlencode(params)}")

    def run_agent_mission(self, **payload: Any) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/agent-missions/run", json=payload)

    def conversations(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/conversations?{urlencode(params)}")

    def conversation(self, conversation_id: str) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/conversations/{conversation_id}")

    def agent_traces(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/agent-traces?{urlencode(params)}")

    def run_reconciliation(self, **payload: Any) -> Dict[str, Any]:
        return self._request("POST", "/v1/angora/reconciliation/run", json=payload)

    def reconciliation_runs(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/reconciliation/runs?{urlencode(params)}")

    def payment_intents(self, **params: Any) -> Dict[str, Any]:
        return self._request("GET", f"/v1/angora/payment-intents?{urlencode(params)}")

File: sdk/typescript/angorapay.ts
export type AngoraSource = "canteen" | "arc-discord" | "circle" | "linkedin" | "x" | "hackathon" | "manual";

export const ANGORAPAY_SDK_VERSION = "0.1.0";

export interface AngoraClientOptions {
  baseUrl?: string;
  gatewayUrl?: string;
  apiKey?: string;
  fetchImpl?: typeof fetch;
}

export interface MarketMissionRequest {
  agentId: string;
  consumerId: string;
  objective: string;
  rfpTrack?: string;
  asset?: string;
  marketContext?: string;
  budget?: string;
  allowedCategories?: string[];
  minProviderTrustScore?: number;
  requiresReceipts?: boolean;
}

export interface GatewayCallRequest {
  missionId: string;
  agentId: string;
  consumerId: string;
  intent: string;
  category: string;
  maxPrice: string;
  payload?: Record<string, unknown>;
  idempotencyKey?: string;
}

export class AngoraPay {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: AngoraClientOptions) {
    const baseUrl = options.baseUrl || options.gatewayUrl;
    if (!baseUrl) throw new Error("AngoraPay requires baseUrl or gatewayUrl");
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = options.apiKey;
    this.fetchImpl = options.fetchImpl || fetch;
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers || {});
    headers.set("Content-Type", "application/json");
    if (this.apiKey) headers.set("Authorization", `Bearer ${this.apiKey}`);

    const response = await this.fetchImpl(`${this.baseUrl}${path}`, { ...init, headers });
    const body = await response.json().catch(() => ({}));
    if (!response.ok) {
      const message = body?.error || `Angora request failed with ${response.status}`;
      throw new Error(message);
    }
    return body as T;
  }

  createMission(input: MarketMissionRequest) {
    return this.request("/v1/angora/missions", { method: "POST", body: JSON.stringify(input) });
  }

  discoverMarketServices(params: { missionId?: string; category?: string; maxPrice?: string; rfpTrack?: string } = {}) {
    const query = new URLSearchParams();
    if (params.missionId) query.set("mission_id", params.missionId);
    if (params.category) query.set("category", params.category);
    if (params.maxPrice) query.set("max_price", params.maxPrice);
    if (params.rfpTrack) query.set("rfp_track", params.rfpTrack);
    return this.request(`/v1/angora/services/search?${query.toString()}`);
  }

  simulateRoute(missionId?: string, maxPrice = "0.01") {
    const query = new URLSearchParams();
    if (missionId) query.set("mission_id", missionId);
    query.set("max_price", maxPrice);
    return this.request(`/v1/angora/route/simulate?${query.toString()}`);
  }

  callMarketService(input: GatewayCallRequest) {
    return this.request("/v1/angora/gateway/call", { method: "POST", body: JSON.stringify(input) });
  }

  runDemoMarketMission(payload: Record<string, unknown> = {}) {
    return this.request("/v1/angora/demo/market-mission", { method: "POST", body: JSON.stringify({ payload }) });
  }

  recordUser(input: { userId?: string; displayName?: string; email?: string; walletAddress?: string; source?: AngoraSource; metadata?: Record<string, unknown> }) {
    return this.request("/v1/angora/traction/users", { method: "POST", body: JSON.stringify(input) });
  }

  recordFeedback(input: { userId: string; missionId?: string; rating: number; comment: string }) {
    return this.request("/v1/angora/traction/feedback", { method: "POST", body: JSON.stringify(input) });
  }

  getExecutionHistory(params: { missionId?: string; status?: string; agentId?: string; providerId?: string } = {}) {
    const query = new URLSearchParams();
    if (params.missionId) query.set("mission_id", params.missionId);
    if (params.status) query.set("status", params.status);
    if (params.agentId) query.set("agent_id", params.agentId);
    if (params.providerId) query.set("provider_id", params.providerId);
    return this.request(`/v1/angora/execution-history?${query.toString()}`);
  }

  getSubmissionMetrics() {
    return this.request("/v1/angora/submission/metrics");
  }

  getProductionReadiness() {
    return this.request("/v1/angora/production/readiness");
  }

  listReceipts(params: { missionId?: string; status?: string; executionMode?: string; limit?: number; offset?: number } = {}) {
    const query = new URLSearchParams();
    if (params.missionId) query.set("mission_id", params.missionId);
    if (params.status) query.set("status", params.status);
    if (params.executionMode) query.set("execution_mode", params.executionMode);
    if (params.limit != null) query.set("limit", String(params.limit));
    if (params.offset != null) query.set("offset", String(params.offset));
    return this.request(`/v1/angora/receipts?${query.toString()}`);
  }

  runAgentMission(input: { userGoal: string; module?: "prediction_market" | "cross_venue_arbitrage" | "social_trading"; conversationId?: string; userId?: string; budgetUSDC?: string; paymentMode?: "real_x402" | "arc_testnet" | "demo_fallback"; payload?: Record<string, unknown> }) {
    return this.request("/v1/angora/agent-missions/run", { method: "POST", body: JSON.stringify(input) });
  }

  listConversations(params: { userId?: string; limit?: number; offset?: number } = {}) {
    const query = new URLSearchParams();
    if (params.userId) query.set("user_id", params.userId);
    if (params.limit != null) query.set("limit", String(params.limit));
    if (params.offset != null) query.set("offset", String(params.offset));
    return this.request(`/v1/angora/conversations?${query.toString()}`);
  }

  getConversation(conversationId: string) {
    return this.request(`/v1/angora/conversations/${conversationId}`);
  }

  getAgentTraces(params: { conversationId?: string; missionId?: string; limit?: number; offset?: number } = {}) {
    const query = new URLSearchParams();
    if (params.conversationId) query.set("conversation_id", params.conversationId);
    if (params.missionId) query.set("mission_id", params.missionId);
    if (params.limit != null) query.set("limit", String(params.limit));
    if (params.offset != null) query.set("offset", String(params.offset));
    return this.request(`/v1/angora/agent-traces?${query.toString()}`);
  }

  runReconciliation(input: { missionId?: string; receiptId?: string } = {}) {
    return this.request("/v1/angora/reconciliation/run", { method: "POST", body: JSON.stringify(input) });
  }

  listReconciliationRuns(params: { missionId?: string; receiptId?: string; limit?: number; offset?: number } = {}) {
    const query = new URLSearchParams();
    if (params.missionId) query.set("mission_id", params.missionId);
    if (params.receiptId) query.set("receipt_id", params.receiptId);
    if (params.limit != null) query.set("limit", String(params.limit));
    if (params.offset != null) query.set("offset", String(params.offset));
    return this.request(`/v1/angora/reconciliation/runs?${query.toString()}`);
  }

  listPaymentIntents(params: { missionId?: string; receiptId?: string; status?: string; limit?: number; offset?: number } = {}) {
    const query = new URLSearchParams();
    if (params.missionId) query.set("mission_id", params.missionId);
    if (params.receiptId) query.set("receipt_id", params.receiptId);
    if (params.status) query.set("status", params.status);
    if (params.limit != null) query.set("limit", String(params.limit));
    if (params.offset != null) query.set("offset", String(params.offset));
    return this.request(`/v1/angora/payment-intents?${query.toString()}`);
  }
}

